File: 2021/7/main.cpp
#include <algorithm>
#include <charconv>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <optional>
#include <string_view>
#include <vector>

std::vector<int> parse_input(std::string_view filename) {
    std::ifstream ifs {filename.data()};

    std::vector<int> out{};
    std::string str_num;
    while(std::getline(ifs, str_num, ','))
    {
        int num;
        std::from_chars(str_num.data(), str_num.data() + str_num.size(), num);
        out.push_back(num);
    }

  return out;
}

template <typename COST_FUNCTION>
void min_alignment_cost(const std::vector<int>& input, COST_FUNCTION&& cost)
{
    auto max_iter = std::max_element(input.begin(), input.end());
    std::optional<std::pair<size_t, int>> min_cost = std::nullopt;

    for(size_t i = 0; i < *max_iter; ++i)
    {
        int cost_count = 0;
        for(const auto num : input)
        {
            cost_count += cost(std::abs(static_cast<int>(num - i)));
        }

        if(!min_cost || (min_cost && std::get<1>(*min_cost) > cost_count))
        {
            min_cost = {i, cost_count};
        }
    }

    std::cout << "Least cost is " << std::get<1>(*min_cost) << " at " << std::get<0>(*min_cost) << '\n';
}

int main() {
    std::vector<int> input = parse_input("in.txt");

    min_alignment_cost(input, [](int dist) { return dist; });
    min_alignment_cost(input, [](int dist) { return (dist * (dist + 1)) / 2; });
}

File: 2021/7/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Error};

fn parse_input(filename: &str) -> Result<Vec<i64>, Error> {
    let reader = BufReader::new(File::open(filename)?);
    let out = reader
        .lines().next().unwrap().unwrap()
        .split(',').map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .map(|s| s.parse().unwrap())
        .collect();
    return Ok(out);
}

fn min_alignment_cost(input: &[i64], cost: &dyn Fn(i64) -> i64) {
    let max_iter = input.iter().max().unwrap();
    let mut min_cost: Option<(i64, i64)> = None;

    for i in 0..*max_iter {
        // Get cost for the movement
        let mut cost_count = 0;
        for num in input.iter() {
            cost_count += cost((*num - i).abs());
        }

        // Check if calculated cost is lowest
        if (min_cost.is_some() && min_cost.unwrap().1 > cost_count) || min_cost.is_none() {
            min_cost = Some((i, cost_count));
        }
    }

    println!("Lowest cost is {} at {}", min_cost.unwrap().1, min_cost.unwrap().0);
}

fn main() {
    let input = parse_input("in.txt")
        .expect("Failed to parse input");

    min_alignment_cost(&input, &|cost| cost);
    min_alignment_cost(&input, &|cost| (cost * (cost + 1)) / 2);
}
